hydroflow/examples/rga/adjacency.rs

```diff
@@ -0,0 +1,96 @@
+        // siblings(bigger, smaller)
+        siblings -> nextSiblingAnc;
+        // lastChild(parent, lastChild)
+        lastChild = adjacency[lastChild] -> map(|(parent, kids): (Timestamp, VecDeque<Timestamp>)| (parent, *kids.front().unwrap()));
+        upEdge = join() -> map(|(_parent, (last_child, next_sib)): (Timestamp, (Timestamp, Timestamp))| (last_child, next_sib));
+        lastChild -> [0]upEdge;
+        nextSiblingAnc -> [1]upEdge;
+        upEdge -> nextSiblingAnc;
+
+        // nextElem
+        nextElem = merge();
+        adjacency[nextElem] -> map(|(parent, kids): (Timestamp, VecDeque<Timestamp>)| (parent, *kids.back().unwrap())) -> nextElem;
+        // nextElem(Prev,Next) :- isListElem(Prev), !hasChild (Prev), nextSiblingAnc(Prev,Next)
+        ne_join = join() -> map(|(prev, (_, next))| (prev, next));
+        leaves -> map(|prev| (prev, ())) -> [0]ne_join;
+        nextSiblingAnc -> [1]ne_join;
+        ne_join -> nextElem;
+
+        insertAfter[print] -> for_each(|(c, p): (Token, Timestamp)| rga_send.send((c, p)).unwrap());
+        nextElem -> for_each(|(first, second)| list_send.send((first, second)).unwrap());
+    }
+}
+
+// // stack subroutine:
+// // - accepts vecDeque batches to insert
+// // - pops one element per tick
+// // each batch is pushed onto the stack in sort order
+// // so that pops are in the correct order (largest id first)
+// stack_buf = merge() -> reduce(|mut accum: VecDeque<Timestamp>, mut v: VecDeque<Timestamp>| {
+//     accum.append(&mut v);
+//     accum
+// });
+// stack = merge()
+//         -> fold(VecDeque::new(), |mut accum: VecDeque<Timestamp>, mut v: VecDeque<Timestamp>| {
+//             v.make_contiguous().sort();
+//             v.iter().for_each(|e| accum.push_front(*e));
+//             accum
+//         });
+// // stack -> for_each(|v: VecDeque<Timestamp>| println!("stack: {:?}", v));
+// stack_buf -> next_tick() -> stack;
+// pop_stack = stack -> map(|mut q: VecDeque<Timestamp>| { let e = q.pop_front().unwrap(); (q, e)}) -> unzip();
+// pop_stack[0] -> filter(|v: &VecDeque<Timestamp>| v.len() > 0) -> stack_buf;
+// visit = pop_stack[1] -> tee();
+
+// // initialize output by pushing root id on the stack
+//         root = insertAfter -> map(|(_child, parent)| parent) -> reduce(|accum: Timestamp, elem: Timestamp| if accum < elem {accum} else {elem} )
+// root -> map(|v| { let mut d = VecDeque::new(); d.push_front(v); d}) -> stack;
+
+// // "visit" a node by looking up its children and pushing them onto the stack
+// children = join::<'tick, 'static>() -> map(|(_n, (_, kids))| kids);
+// visit -> map(|n| (n, ())) -> [0]children;
+// adjacency -> [1]children;
+// children -> stack_buf;
+
+// labels = insertAfter -> map(|(c, _p): (Token, Timestamp)| (c.id, c.value));
+// visit_labels = join::<'tick, 'static>() -> map(|(_id, ((), label))| label);
+// visit -> map(|n| (n, ())) -> [0]visit_labels;
+// labels -> [1]visit_labels;
+// visit_labels -> for_each(|e| print!("{}", e));
```

Comment:
commented out code?

Reply:
OK I'll keep this somewhere else.

Reply:
Oh its fine to keep here, just not sure what it is for

Reply:
it's an implementation of a stack that pops one item per tick. I'll save it on the side -- something we should discuss.